### DISCUSSION TITLE
Show modal dialog window from WinFormsAvaloniaControlHost

### DISCUSSION THREAD

**Discussion by pauldumais** [AvaloniaUI/Avalonia#15977] [Q&A] [answered]
We are in the process of converting our large WinForms application to Avalonia, we have converted several views using WinFormsAvaloniaControlHost this works great!  

We would like to display some modal avalonia dialog windows from the WinFormsAvaloniaControlHost avalonia view, for example when the user clicks a Button.  Is this possible?

We have tried access Avalonia.Application.Current.ApplicationLifetime but this value appears to be null when running inside a WinForms app.

Example:

```
            var lifetime = Avalonia.Application.Current.ApplicationLifetime;

            // lifetime is null here

            IClassicDesktopStyleApplicationLifetime desktop = lifetime as IClassicDesktopStyleApplicationLifetime;
            
            var window = new TestWindow1();
            window.ShowDialog(desktop.MainWindow);

```

The only workaround we have found is to continue using WinForms Form dialog objects and embed a WinFormsAvaloniaControlHost inside each dialog, which is a bit messy, it would be nice if we could us the Avalonia Window object to clean this up.

**Comment by thevortexcloud** [comment]
Do you have a real Avalonia window? The winforms interop control runs the Avalonia control inside a winform form, not an Avalonia window. If you have an Avalonia window you can just make it display a dialogue directly. You should also be able to instantiate an Avalonia window yourself as well. Although I am not too sure how well that works with winforms.

**Comment by maxkatz6** [comment]
I don't think it's possible with the current public API.
But I can imagine following to work well:
1. Create an Avalonia Window.
2. Get HWND of avalonia window (.TryGetPlatformHandle()) and of winforms window.
3. Show the window.
4. Call WinAPI to set owner/parent of the child window after its shown.

"After its shown" is likely to be necessary, otherwise Avalonia would redefine it with null-parent.

**Reply by maxkatz6** [reply]
We might be able to add an extension method to show window on HWND parent. But it would be more difficult to get passed.

**Reply by pauldumais** [reply, marked as answer]
Thanks @maxkatz6 .  I was able to get it to work following your instructions, I had to make a few changes, here is the code that worked for me to show a modal Avalonia Window from a WinForm app:

```
    public static void ShowModalWindow(System.Windows.Forms.Form winForm, Avalonia.Controls.Window avaloniaWindow)
    {
        var parentHandle = winForm.Handle;
        var dialogHandle = avaloniaWindow.TryGetPlatformHandle().Handle;
        SetWindowLong(dialogHandle, GWLP_HWNDPARENT, parentHandle);

        avaloniaWindow.Show();
        avaloniaWindow.Closed += (sender, args) => 
            {
                EnableWindow(parentHandle, true);
                SetWindowPos(parentHandle, HWND_TOP, 0, 0, 0, 0, SWP_NOMOVE|SWP_NOSIZE);
            };
        EnableWindow(parentHandle, false);
    }
```
